gnovm/pkg/gnomod/pkg.go

```diff
@@ -0,0 +1,142 @@
+package gnomod
+
+import (
+	"fmt"
+	"io/fs"
+	"os"
+	"path/filepath"
+)
+
+type pkg struct {
+	name     string
+	path     string
+	requires []string
+}
+
+// Name returns the name of the package.
+func (p pkg) Name() string {
+	return p.name
+}
+
+// Path returns the path of the package.
+func (p pkg) Path() string {
+	return p.path
+}
+
+// Requires returns the required packages of the package.
+func (p pkg) Requires() []string {
+	return p.requires
+}
+
+// sortPkgs sorts the given packages by their dependencies.
+func SortPkgs(pkgs []pkg) error {
+	visited := make(map[string]bool)
+	onStack := make(map[string]bool)
+	sortedPkgs := make([]pkg, 0, len(pkgs))
+
+	// Visit all packages
+	for _, p := range pkgs {
+		if err := visitPackage(p, pkgs, visited, onStack, &sortedPkgs); err != nil {
+			return err
+		}
+	}
+
+	copy(pkgs, sortedPkgs)
+	return nil
+}
+
+// visitNode visits a package's and its dependencies dependencies and adds them to the sorted list.
+func visitPackage(pkg pkg, pkgs []pkg, visited, onStack map[string]bool, sortedPkgs *[]pkg) error {
+	if onStack[pkg.name] {
+		return fmt.Errorf("cycle detected: %s", pkg.name)
+	}
+	if visited[pkg.name] {
+		return nil
+	}
+
+	visited[pkg.name] = true
+	onStack[pkg.name] = true
+
+	// Visit package's dependencies
+	for _, req := range pkg.requires {
+		found := false
+		for _, p := range pkgs {
+			if p.name != req {
+				continue
+			}
+			if err := visitPackage(p, pkgs, visited, onStack, sortedPkgs); err != nil {
+				return err
+			}
+			found = true
+			break
+		}
+		if !found {
+			return fmt.Errorf("missing dependency '%s' for package '%s'", req, pkg.name)
+		}
+	}
+
+	onStack[pkg.name] = false
+	*sortedPkgs = append(*sortedPkgs, pkg)
+	return nil
+}
+
+// ListPkgs lists all gno packages in the given root directory.
+func ListPkgs(root string) ([]pkg, error) {
+	var pkgs []pkg
+	draft := make(map[string]bool)
+
+	err := filepath.WalkDir(root, func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+		if !d.IsDir() {
+			return nil
+		}
+		goModPath := filepath.Join(path, "gno.mod")
+		data, err := os.ReadFile(goModPath)
+		if os.IsNotExist(err) {
+			return nil
+		}
+		if err != nil {
+			return err
+		}
+
+		gnoMod, err := Parse(goModPath, data)
+		if err != nil {
+			return fmt.Errorf("parse: %w", err)
+		}
+		gnoMod.Sanitize()
+		if err := gnoMod.Validate(); err != nil {
+			return fmt.Errorf("validate: %w", err)
+		}
+
+		// Ignore if draft or depends on draft
+		if gnoMod.Draft {
+			draft[gnoMod.Module.Mod.Path] = true
+			return fs.SkipDir
+		}
+		for _, req := range gnoMod.Require {
+			if draft[req.Mod.Path] {
+				return fs.SkipDir
+			}
+		}
```

Comment:
@harry-hov Shouldn't you do the same thing for `gnoMod.Replace` ?

Something like 

```go
		for _, rep := range gnoMod.Replace {
			if draft[rep.New.Path] {
				return fs.SkipDir
			}
		}
```

Reply:
It would be nice to test this function. Whether by using a `testdata` folder but that's tedious, or better by using the awesome [`testscript`][0] library :)

[0]: https://pkg.go.dev/github.com/rogpeppe/go-internal/testscript

Reply:
Umm. I don't think it's good idea to check `replace` directive for draft pkgs. 
Ultimately the vm or chain is not going to obey the `replace` directive.

(right now replace is for testing purpose only)

Reply:
Actually I was thinking about writing test for this function. Thanks for suggesting  [`testscript`][0]. I'll check it out.

[0]: https://pkg.go.dev/github.com/rogpeppe/go-internal/testscript

Reply:
Didn't used  [`testscript`](https://pkg.go.dev/github.com/rogpeppe/go-internal/testscript) as I don't want to add new dep. 

See 9282dbab4fa6b45179302e41f8ab7ab64a4a3dbd